File: ios/Sources/TaigiKeyboard/Settings/BuiltInThemes.swift
// Static built-in theme table: three key-style families over one shared set of 7 colors.
// Transparent keys (框線 / 簡潔) set normalKeyFill/specialKeyFill to clear so the keyboard background
// shows through. The 5 gradients are light-only, 暗眠山貓 is dark-only, 預設 stays adaptive (bg/text
// left nil so the renderer resolves them from the system colorScheme).

import Foundation

/// One built-in theme family — a section header (`title`) plus its variant
/// themes, shown as a horizontal shelf in the theme picker. The three families
/// (經典 / 框線 / 簡潔) are a key-STYLE axis over one shared set of 7 colors.
struct BuiltInThemeFamily: Equatable {
    let titleKey: StringKey
    let themes: [BuiltInTheme]
}

/// The app-bundled, read-only theme catalog. Three key-style families, each with
/// the same 7 colors (adaptive 預設 + 5 light-only gradients + 1 dark-only 暗眠山貓).
/// 經典 keeps filled keys; 框線 / 簡潔 make keys transparent (background shows
/// through), 框線 adding an outline.
enum BuiltInThemes {
    /// Shelf order shown to the user (after `Default`, before user themes).
    static let families: [BuiltInThemeFamily] = [
        BuiltInThemeFamily(titleKey: .themeFamilyClassic, themes: familyThemes(.classic)),
        BuiltInThemeFamily(titleKey: .themeFamilyFramed, themes: familyThemes(.framed)),
        BuiltInThemeFamily(titleKey: .themeFamilyClean, themes: familyThemes(.clean)),
    ]

    /// Flattened lookup roster — used by `ThemeResolver` / `SharedSettings` to
    /// resolve a selected id.
    static let all: [BuiltInTheme] = families.flatMap(\.themes)

    /// Looks up a built-in by id; `nil` when the id is not a built-in.
    static func theme(id: String) -> BuiltInTheme? {
        all.first { $0.id == id }
    }

    // MARK: - Key style

    /// The per-family key-style axis. All three families share one set of colors;
    /// only the key rendering differs.
    private enum KeyStyle {
        case classic // filled keys (white over a gradient, adaptive for 預設)
        case framed // transparent keys + outline border
        case clean // transparent keys, no border

        /// Keys are transparent (background shows through) for framed / clean.
        var hasTransparentKeys: Bool { self != .classic }
        /// Only the framed family draws the key outline.
        var isBordered: Bool { self == .framed }
        /// id prefix per family — 經典 keeps the legacy `standard*` ids.
        var idPrefix: String {
            switch self {
            case .classic: "standard"
            case .framed: "framed"
            case .clean: "clean"
            }
        }
    }

    // MARK: - Shared colors

    /// One of the 7 shared color identities. `gradient == nil` is the adaptive
    /// 預設 head; the next 5 are soft light single-hue gradients; the last is the
    /// dark-only 暗眠山貓 (Catppuccin Mocha) gradient (`isDarkPalette`), which lands
    /// in the `dark` variant slot with light text over a dark gradient.
    private struct BaseColor {
        let key: String
        let displayNameKey: StringKey
        let gradient: (top: UInt32, bottom: UInt32)?
        /// Dark palette (e.g. 暗眠山貓/Catppuccin): light text over a dark gradient,
        /// with a dark neutral key fill; builds into the `dark` variant slot
        /// (`light` nil) so it stays dark regardless of system scheme.
        var isDarkPalette: Bool = false
    }

    // Gradient hex values are visual approximations.
    private static let baseColors: [BaseColor] = [
        BaseColor(key: "default", displayNameKey: .themePaletteDefault, gradient: nil),
        BaseColor(key: "pink", displayNameKey: .themePalettePink, gradient: (0xE6C2D0, 0xEADCE2)),
        BaseColor(key: "gold", displayNameKey: .themePaletteGold, gradient: (0xEAD9A6, 0xECE4D2)),
        BaseColor(key: "blue", displayNameKey: .themePaletteBlue, gradient: (0xBFD2EA, 0xDCE2EC)),
        BaseColor(key: "green", displayNameKey: .themePaletteGreen, gradient: (0xC3D8C8, 0xDCE5DD)),
        BaseColor(key: "purple", displayNameKey: .themePalettePurple, gradient: (0xCDC4E4, 0xDEDAEA)),
        // 暗眠山貓 = Catppuccin Mocha: Base→Mantle background (darker than keys), Surface0 keys, Text glyphs.
        BaseColor(key: "catppuccin", displayNameKey: .themePaletteCatppuccin, gradient: (0x1E1E2E, 0x181825), isDarkPalette: true),
    ]

    // Neutral key fill + text for gradient themes — a light pair and a dark (Catppuccin Mocha) pair.
    // CROSS-PLATFORM INVARIANT — mirrors android .../ime/core/BuiltInThemes.kt LIGHT_KEY_FILL/LIGHT_KEY_TEXT/DARK_KEY_FILL/DARK_KEY_TEXT.
    // Drift causes silent divergence (iOS/Android theme key colors differ).
    private static let lightKeyFill: UInt32 = 0xFFFFFF
    private static let lightKeyText: UInt32 = 0x1C1C1E
    private static let darkKeyFill: UInt32 = 0x313244 // Catppuccin Mocha Surface0
    private static let darkKeyText: UInt32 = 0xCDD6F4 // Catppuccin Mocha Text

    // CROSS-PLATFORM INVARIANT — mirrors android .../ime/core/BuiltInThemes.kt OUTLINED_KEY_BORDER_WIDTH.
    // Drift causes silent divergence.
    private static let outlinedKeyBorderWidth: Double = 1.0

    // MARK: - Builders

    /// Builds the 7 themes for one key-style family. The 經典 head keeps the
    /// `ThemeId.default` sentinel (so reset shows it selected); framed / clean use
    /// `framedDefault` / `cleanDefault` ids. A light theme builds into the `light`
    /// slot (`dark` nil); a dark theme (暗眠山貓) builds into the `dark` slot
    /// (`light` nil) — mirror-symmetric. Preview slots mirror the family id prefix;
    /// missing assets fall back to a neutral placeholder until screenshots ship.
    private static func familyThemes(_ style: KeyStyle) -> [BuiltInTheme] {
        baseColors.map { base in
            let isDefault = base.gradient == nil
            let suffix = base.key.prefix(1).uppercased() + base.key.dropFirst()
            let id = isDefault
                ? (style == .classic ? ThemeId.default : "\(style.idPrefix)Default")
                : "\(style.idPrefix)\(suffix)"
            let previewName = isDefault
                ? "theme_\(style.idPrefix)_preview"
                : "theme_\(style.idPrefix)\(suffix)_preview"
            let scheme = colors(for: base, style: style)
            return BuiltInTheme(
                id: id,
                displayNameKey: base.displayNameKey,
                light: base.isDarkPalette ? nil : scheme,
                dark: base.isDarkPalette ? scheme : nil,
                previewImageName: previewName,
                keyBorderWidth: style.isBordered ? outlinedKeyBorderWidth : nil,
            )
        }
    }

    /// Resolves the color palette for one (color, key-style) pair. 經典 預設 stays
    /// fully adaptive (`nil`); framed / clean 預設 carry only transparent key fills
    /// so the adaptive background/text still show through and adapt to dark mode.
    private static func colors(for base: BaseColor, style: KeyStyle) -> KeyboardColorSettings? {
        if let gradient = base.gradient {
            let keyText = base.isDarkPalette ? darkKeyText : lightKeyText
            let neutralFill = base.isDarkPalette ? darkKeyFill : lightKeyFill
            return gradientColors(top: gradient.top, bottom: gradient.bottom, keyText: keyText, neutralFill: neutralFill, transparentKeys: style.hasTransparentKeys)
        }
        guard style.hasTransparentKeys else { return nil } // 經典 預設 = adaptive
        var colors = KeyboardColorSettings()
        colors.normalKeyFillColor = CodableColor(.clear)
        colors.specialKeyFillColor = CodableColor(.clear)
        return colors
    }

    /// One scheme variant for a gradient color: the 2-stop background gradient +
    /// key/candidate text (`keyText`). Keys are either the neutral fill (`neutralFill`,
    /// 經典) or transparent so the gradient shows through (框線 / 簡潔). `backgroundColor`
    /// and `candidateBackgroundColor` stay nil — the gradient owns the background
    /// and the candidate bar is made transparent in `TaigiKeyboardView.candidateStyle`.
    private static func gradientColors(top: UInt32, bottom: UInt32, keyText: UInt32, neutralFill: UInt32, transparentKeys: Bool) -> KeyboardColorSettings {
        var colors = KeyboardColorSettings()
        colors.backgroundGradient = ThemeGradient(stops: [CodableColor(hex: top), CodableColor(hex: bottom)])
        colors.keyTextColor = CodableColor(hex: keyText)
        colors.candidateTextColor = CodableColor(hex: keyText)
        if transparentKeys {
            colors.normalKeyFillColor = CodableColor(.clear)
            colors.specialKeyFillColor = CodableColor(.clear)
        } else {
            let fill = CodableColor(hex: neutralFill)
            colors.normalKeyFillColor = fill
            colors.specialKeyFillColor = fill
        }
        return colors
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/core/BuiltInThemes.kt
// Built-in theme catalog — three key-style families over one shared 7-color palette (iOS parity).

package com.siansiansu.taigikeyboard.ime.core

import com.siansiansu.taigikeyboard.i18n.generated.StringKey

/**
 * One built-in, read-only theme: a named palette with optional light/dark 6-role
 * color variants resolved against night-mode at render time. A null variant falls
 * back to the other. A theme may also carry one appearance override
 * ([keyBorderWidth], used by the 框線 family). Mirrors iOS BuiltInTheme.
 */
data class BuiltInTheme(
    val id: String,
    // i18n key for the display name, resolved at the picker call site via the
    // active StringResolver so the name follows the user's chosen display language.
    val displayNameKey: StringKey,
    val light: KeyboardColorSettings?,
    val dark: KeyboardColorSettings?,
    val previewImageName: String? = null,
    val keyBorderWidth: Float? = null,
) {
    /** Picks the variant for [isDark], falling back to the other when absent. */
    fun colors(isDark: Boolean): KeyboardColorSettings =
        if (isDark) {
            dark ?: light ?: KeyboardColorSettings()
        } else {
            light ?: dark ?: KeyboardColorSettings()
        }
}

/**
 * One built-in family — a section header + its variant themes (one picker shelf).
 * The three families (經典 / 框線 / 簡潔) are a key-STYLE axis over one shared set
 * of 7 colors.
 */
data class BuiltInThemeFamily(
    val titleKey: StringKey,
    val themes: List<BuiltInTheme>,
)

/**
 * The app-bundled, read-only theme catalog. Three key-style families, each with
 * the same 7 colors (adaptive 預設 + 5 light-only gradients + 1 dark-only 暗眠山貓).
 * 經典 keeps filled keys; 框線 / 簡潔 make keys transparent (background shows
 * through), 框線 adding an outline. Mirrors iOS BuiltInThemes.
 */
object BuiltInThemes {
    // `by lazy` so the catalog builds on FIRST ACCESS, not during object init: a
    // plain `val` here would run `familyThemes` (which reads the `baseColors` val
    // declared below) before this object finishes constructing top-to-bottom, so
    // `baseColors` would still be null. iOS gets this for free (`static let` is lazy).
    val families: List<BuiltInThemeFamily> by lazy {
        listOf(
            BuiltInThemeFamily(StringKey.THEME_FAMILY_CLASSIC, familyThemes(KeyStyle.CLASSIC)),
            BuiltInThemeFamily(StringKey.THEME_FAMILY_FRAMED, familyThemes(KeyStyle.FRAMED)),
            BuiltInThemeFamily(StringKey.THEME_FAMILY_CLEAN, familyThemes(KeyStyle.CLEAN)),
        )
    }

    /** Flattened lookup roster — used by [ThemeResolver] to resolve a selected id. */
    val all: List<BuiltInTheme> by lazy { families.flatMap { it.themes } }

    /** Looks up a built-in by id; null when [id] is not a built-in. */
    fun theme(id: String): BuiltInTheme? = all.firstOrNull { it.id == id }

    /**
     * The per-family key-style axis. All three families share one set of colors;
     * only the key rendering differs. [idPrefix] keeps 經典 on the legacy
     * `standard*` ids.
     */
    private enum class KeyStyle(val idPrefix: String) {
        CLASSIC("standard"), // filled keys (white over a gradient, adaptive for 預設)
        FRAMED("framed"), // transparent keys + outline border
        CLEAN("clean"), // transparent keys, no border
        ;

        val hasTransparentKeys: Boolean get() = this != CLASSIC
        val isBordered: Boolean get() = this == FRAMED
    }

    /**
     * One of the 7 shared color identities. [gradient] == null is the adaptive 預設
     * head; the next 5 are soft light single-hue gradients (raw 0xRRGGBB pairs); the
     * last is the dark-only 暗眠山貓 (Catppuccin Mocha) gradient ([isDarkPalette]), which
     * lands in the `dark` variant slot with light text over a dark gradient.
     */
    private data class BaseColor(
        val key: String,
        val displayNameKey: StringKey,
        val gradient: Pair<Int, Int>?,
        val isDarkPalette: Boolean = false,
    )

    // Order matches the picker shelf; gradient hex values are visual approximations.
    private val baseColors: List<BaseColor> =
        listOf(
            BaseColor("default", StringKey.THEME_PALETTE_DEFAULT, null),
            BaseColor("pink", StringKey.THEME_PALETTE_PINK, 0xE6C2D0 to 0xEADCE2),
            BaseColor("gold", StringKey.THEME_PALETTE_GOLD, 0xEAD9A6 to 0xECE4D2),
            BaseColor("blue", StringKey.THEME_PALETTE_BLUE, 0xBFD2EA to 0xDCE2EC),
            BaseColor("green", StringKey.THEME_PALETTE_GREEN, 0xC3D8C8 to 0xDCE5DD),
            BaseColor("purple", StringKey.THEME_PALETTE_PURPLE, 0xCDC4E4 to 0xDEDAEA),
            // Catppuccin Mocha Base→Mantle background gradient, deliberately darker than the keys.
            BaseColor("catppuccin", StringKey.THEME_PALETTE_CATPPUCCIN, 0x1E1E2E to 0x181825, isDarkPalette = true),
        )

    // CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/BuiltInThemes.swift lightKeyFill/lightKeyText/darkKeyFill/darkKeyText.
    // Drift causes silent divergence (iOS/Android theme key colors differ).
    private const val LIGHT_KEY_FILL = 0xFFFFFF
    private const val LIGHT_KEY_TEXT = 0x1C1C1E
    private const val DARK_KEY_FILL = 0x313244 // Catppuccin Mocha Surface0
    private const val DARK_KEY_TEXT = 0xCDD6F4 // Catppuccin Mocha Text

    // Transparent fill for the 框線 / 簡潔 families — the keyboard background shows through.
    private const val TRANSPARENT_KEY_FILL = 0x00000000

    // CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Settings/BuiltInThemes.swift outlinedKeyBorderWidth.
    // Drift causes silent divergence.
    private const val OUTLINED_KEY_BORDER_WIDTH = 1.0f

    /**
     * Builds the 7 themes for one key-style family. The 經典 head keeps the
     * [ThemeId.DEFAULT] sentinel (so reset shows it selected); framed / clean use
     * `framedDefault` / `cleanDefault` ids. A light theme builds into the `light`
     * slot (`dark` null); a dark theme (暗眠山貓) builds into the `dark` slot (`light`
     * null) — mirror-symmetric. Preview slots mirror the family id prefix; missing
     * assets fall back to a neutral placeholder until screenshots ship.
     */
    private fun familyThemes(style: KeyStyle): List<BuiltInTheme> =
        baseColors.map { base ->
            val isDefault = base.gradient == null
            val suffix = base.key.replaceFirstChar { it.uppercase() }
            val id =
                when {
                    !isDefault -> "${style.idPrefix}$suffix"
                    style == KeyStyle.CLASSIC -> ThemeId.DEFAULT
                    else -> "${style.idPrefix}Default"
                }
            val previewName =
                if (isDefault) "theme_${style.idPrefix}_preview" else "theme_${style.idPrefix}${suffix}_preview"
            val scheme = colorsFor(base, style)
            BuiltInTheme(
                id = id,
                displayNameKey = base.displayNameKey,
                light = if (base.isDarkPalette) null else scheme,
                dark = if (base.isDarkPalette) scheme else null,
                previewImageName = previewName,
                keyBorderWidth = if (style.isBordered) OUTLINED_KEY_BORDER_WIDTH else null,
            )
        }

    /**
     * Resolves the color palette for one (color, key-style) pair. 經典 預設 stays
     * fully adaptive (null); framed / clean 預設 carry only transparent key fills so
     * the adaptive background/text still show through and adapt to dark mode.
     */
    private fun colorsFor(base: BaseColor, style: KeyStyle): KeyboardColorSettings? {
        base.gradient?.let { (top, bottom) ->
            val keyText = if (base.isDarkPalette) DARK_KEY_TEXT else LIGHT_KEY_TEXT
            val neutralFill = if (base.isDarkPalette) DARK_KEY_FILL else LIGHT_KEY_FILL
            return gradientColors(top, bottom, keyText, neutralFill, style.hasTransparentKeys)
        }
        if (!style.hasTransparentKeys) return null // 經典 預設 stays adaptive
        return KeyboardColorSettings(
            normalKeyFillColor = TRANSPARENT_KEY_FILL,
            specialKeyFillColor = TRANSPARENT_KEY_FILL,
        )
    }

    /**
     * One scheme variant for a gradient color: the 2-stop background gradient +
     * key/candidate text ([keyText]). Keys are either the neutral fill ([neutralFill],
     * 經典) or transparent so the gradient shows through (框線 / 簡潔). backgroundColor
     * and candidateBackgroundColor stay null so the gradient owns the background and
     * the candidate bar is transparent over it. light/dark themes pass their own
     * keyText/neutralFill.
     */
    private fun gradientColors(top: Int, bottom: Int, keyText: Int, neutralFill: Int, transparentKeys: Boolean): KeyboardColorSettings {
        val fill = if (transparentKeys) TRANSPARENT_KEY_FILL else argb(neutralFill)
        return KeyboardColorSettings(
            keyTextColor = argb(keyText),
            normalKeyFillColor = fill,
            specialKeyFillColor = fill,
            candidateTextColor = argb(keyText),
            backgroundGradient = ThemeGradient(listOf(argb(top), argb(bottom))),
        )
    }

    /** 0xRRGGBB -> opaque ARGB Int (alpha forced 0xFF). */
    private fun argb(rgb: Int): Int = (0xFF shl 24) or (rgb and 0xFFFFFF)
}
